### PR TITLE
Add pretty print

### PR DESCRIPTION
**Pull request by asterite**
This adds pretty printing to Crystal objects. And now when you do `p obj` that actually pretty prints `obj` instead of using `inspect`. The reason for this is that I always find Ruby's `pp` better whenever I want to quickly "inspect" an object, so I think pretty printing should be preferred. There's also Elixir, which only has `IO.inspect` which does pretty printing, and that never bothered me.

Example:

```cr
class Point
  def initialize
    @x = 1
    @y = 2
    @z = 3
  end
end

class Foo
  @points : Array(Point)

  def initialize
    @points = [Point.new] * 3
  end
end

class Bar
  def initialize
    @foo = Foo.new
  end
end

bar = Bar.new
p bar
```

Output before:

```
#<Bar:0x103173f00 @foo=#<Foo:0x103173ee0 @points=[#<Point:0x103173ea0 @x=1, @y=2, @z=3>, #<Point:0x103173ea0 @x=1, @y=2, @z=3>, #<Point:0x103173ea0 @x=1, @y=2, @z=3>]>>
```

Output after:

```
#<Bar:0x109f16f00
 @foo=
  #<Foo:0x109f16ee0
   @points=
    [#<Point:0x109f16ea0 @x=1, @y=2, @z=3>,
     #<Point:0x109f16ea0 @x=1, @y=2, @z=3>,
     #<Point:0x109f16ea0 @x=1, @y=2, @z=3>]>>
```

I think I prefer the later to be outputted, always (one can do `puts obj.inspect` to get the old behaviour)

I'll be honest: I read [the](http://homepages.inf.ed.ac.uk/wadler/topics/language-design.html#prettier) [papers](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.34.2200) about pretty printing and did an [implementation](https://gist.github.com/asterite/d1b9b8dd1ed11a5e2018973ab740b778) myself, but then I checked how it was done in Ruby and it was much better. The difference is that Ruby streams directly to the given output whenever possible, while my implementation constructed the whole tree and then outputted. In the end, my implementation consumed a lot of memory... so I decided to first port [Ruby's code](https://github.com/ruby/ruby/blob/trunk/lib/prettyprint.rb) to Crystal, check if it works, and then try to understand it. Not surprisingly, porting the code was extremely easy! And it worked out of the box. I think I understand now how it works, though the algorithm is pretty complex (if somebody would like to add more comments afterwards, please do so!). As always, I don't know if I should do something about the copyright, the algorithms seem to be public though I'm not so sure about Ruby's code.

Now we'll have `to_s`, `inspect` and `pretty_print`... maybe we can remove `inspect`? Or maybe we can declare it such that it creates a pretty printer that will output everything into a single line and then invoke `pretty_print` with it? In that way users would only need to override `to_s` and/or `pretty_print`, but never `inspect`. In any case, maybe that can be decided after we merge this PR.

### CODE REVIEW

**Comment by sdogruyol** [comment]
👍 for removing `inspect`

**Comment by RX14** [comment]
I'm in favour of keeping `inspect`, but making it delegate to `pretty_print` by default.

**Comment by kostya** [comment]
why remove inspect?
this is useful when write some object into log LOG.info "... #{obj.inspect} ...", and pretty_print would bad in log.

**Comment by Sija** [comment]
I'd keep `inspect` too, its output is meant to be more terse than `pretty_print` (for debug purposes) and quite often may be different than `to_s`.

File: spec/std/hash_spec.cr [hunk 1]
```diff
@@ -279,12 +279,12 @@ describe "Hash" do
   end
 
   describe "to_s" do
-    assert { {1 => 2, 3 => 4}.to_s.should eq("{1 => 2, 3 => 4}") }
+    assert { {1 => 2, 3 => 4}.to_s.should eq("{1=>2, 3=>4}") }
```
Comment (luislavena) [inline spec/std/hash_spec.cr]: I'm not sure about this, I would prefer if by default the generated representation is valid and properly formatted Crystal code, so paste the output around doesn't trigger formatter differences?

**Pull request merged**